[WPE] WPE Platform: add a connect method to WPEDisplayDRM to use a custom device file

siwei-li opened the pull request:
#### f98baf30b97d9a1b2abedcf66839b950a959bb50
<pre>
[WPE] WPE Platform: add a connect method to WPEDisplayDRM to use a custom device file
<a href="https://bugs.webkit.org/show_bug.cgi?id=265654">https://bugs.webkit.org/show_bug.cgi?id=265654</a>

Reviewed by Carlos Garcia Campos.

Adding a wpe_display_drm_connect() to the DRM platform&apos;s API, so that
the DRM device to connect to may be specified by the device name.

* Source/WebKit/WPEPlatform/wpe/drm/WPEDisplayDRM.cpp:
(wpeDisplayDRMSetup): a setup function to be called by the two functions
below
(wpeDisplayDRMConnect):
(wpe_display_drm_connect):
* Source/WebKit/WPEPlatform/wpe/drm/WPEDisplayDRM.h:

Canonical link: <a href="https://commits.webkit.org/285545@main">https://commits.webkit.org/285545@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/29dec7bd6c70a9f363147bc1823c59a935e25547

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/72924 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/52355 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/25736 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/77206 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/24162 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/60154 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/23974 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/57376 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/15822 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/75991 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/47333 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/62776 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/37796 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/43977 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/22491 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/65833 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/20599 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/78827 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/17170 "Built successfully") | [⏳ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/macOS-Sequoia-Debug-WK2-Tests-EWS "Waiting to run tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/65802 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/17219 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/62777 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/65105 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/13370 "Build is being retried. Recent messages:OS: Sonoma (14.5), Xcode: 15.4; Skipping applying patch since patch_id isn't provided; Checked out pull request; Running run-layout-tests-in-stress-mode; layout-tests running") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/7039 "Passed tests") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/11232 "Built successfully and passed tests") | [![loading-orange](https://github-production-user-asset-6210df.s3.amazonaws.com/3098702/291015173-08c448be-ac0a-4fd6-92a3-8165057445b7.png) 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/48147 "Build is in progress. Recent messages:OS: Sonoma (14.5), Xcode: 15.4; Skipping applying patch since patch_id isn't provided; Checked out pull request; compiling") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/2940 "Built successfully") | | 
| | [⏳ 🛠 tv-sim ](https://ews-build.webkit.org/#/builders/tvOS-18-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | | | 
| | [⏳ 🛠 watch ](https://ews-build.webkit.org/#/builders/watchOS-11-Build-EWS "Waiting in queue, processing has not started yet") | | | 
| | [⏳ 🛠 watch-sim ](https://ews-build.webkit.org/#/builders/watchOS-11-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | | | 
<!--EWS-Status-Bubble-End-->

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/headless/WPEDisplayHeadless.cpp:
```diff
@@ -219,3 +219,46 @@ WPEDisplay* wpe_display_headless_new(void)
 }
+
+/**
+ * wpe_display_headless_new_for_device:
```
```suggestion
 * wpe_display_headless_new_for_device: (skip)
```

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/headless/WPEDisplayHeadless.cpp:
```diff
@@ -219,3 +219,46 @@ WPEDisplay* wpe_display_headless_new(void)
+            }
+        }
+    }
+
```
I would use a different error here, ` DRM device "%s" not found`, or something similar.

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/headless/WPEDisplayHeadless.cpp:
```diff
@@ -219,3 +219,46 @@ WPEDisplay* wpe_display_headless_new(void)
+
+    drmFreeDevices(devices, numDevices);
+#endif
+    g_set_error(error, WPE_DISPLAY_ERROR, WPE_DISPLAY_ERROR_NOT_SUPPORTED, "DRM device not supported");
```
```suggestion
    g_set_error_literal(error, WPE_DISPLAY_ERROR, WPE_DISPLAY_ERROR_NOT_SUPPORTED, "DRM device not supported");
```

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/headless/WPEDisplayHeadless.cpp:
```diff
@@ -219,3 +219,46 @@ WPEDisplay* wpe_display_headless_new(void)
+
+    int numDevices = drmGetDevices2(0, devices, std::size(devices));
+    if (numDevices <= 0) {
+        g_set_error(error, WPE_DISPLAY_ERROR, WPE_DISPLAY_ERROR_NOT_SUPPORTED, "No DRM devices found");
```
```suggestion
        g_set_error_literal(error, WPE_DISPLAY_ERROR, WPE_DISPLAY_ERROR_NOT_SUPPORTED, "No DRM devices found");
```

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/headless/WPEDisplayHeadless.cpp:
```diff
@@ -219,3 +219,46 @@ WPEDisplay* wpe_display_headless_new(void)
+                if (device->available_nodes & (1 << DRM_NODE_RENDER))
+                    priv->drmRenderNode = CString(device->nodes[DRM_NODE_RENDER]);
+                if (device->available_nodes & (1 << DRM_NODE_PRIMARY))
+                    priv->drmDevice = CString(device->nodes[DRM_NODE_PRIMARY]);
```
If there's no primary or render node, we should initialize to a null CString.

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/drm/WPEDisplayDRM.cpp:
```diff
@@ -413,6 +427,22 @@ WPEDisplay* wpe_display_drm_new(void)
+ *
+ * Returns: %TRUE if connection succeeded, or %FALSE in case of error.
+ */
+gboolean wpe_display_drm_connect(WPEDisplayDRM* display, const char* name, GError** error)
+{
```
Add `g_return_val_if_fail(WPE_IS_DISPLAY_DRM(display), FALSE);`

siwei-li committed:
[WPE] WPE Platform: add a connect method to WPEDisplayDRM to use a custom device file
https://bugs.webkit.org/show_bug.cgi?id=265654

Reviewed by Carlos Garcia Campos.

Adding a wpe_display_drm_connect() to the DRM platform's API, so that
the DRM device to connect to may be specified by the device name.

* Source/WebKit/WPEPlatform/wpe/drm/WPEDisplayDRM.cpp:
(wpeDisplayDRMSetup): a setup function to be called by the two functions
below
(wpeDisplayDRMConnect):
(wpe_display_drm_connect):
* Source/WebKit/WPEPlatform/wpe/drm/WPEDisplayDRM.h:

Canonical link: https://commits.webkit.org/285545@main
